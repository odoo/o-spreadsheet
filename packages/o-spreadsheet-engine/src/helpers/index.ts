export * from "./misc";
export * from "./numbers";
export * from "./references";
export * from "./uuid";
