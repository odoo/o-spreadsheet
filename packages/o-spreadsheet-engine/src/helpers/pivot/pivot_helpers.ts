import { boolAnd, boolOr } from "../../functions/helper_logical";
import { countUnique, sum } from "../../functions/helper_math";
import { average, countAny, max, min } from "../../functions/helper_statistical";
import {
  inferFormat,
  isEvaluationError,
  toBoolean,
  toNumber,
  toString,
  toValue,
} from "../../functions/helpers";
import { Registry } from "../../registries/registry";
import { _t } from "../../translation";
import { CellValue } from "../../types/cells";
import { EvaluationError } from "../../types/errors";
import { DEFAULT_LOCALE, Locale } from "../../types/locale";
import { CellPosition, FunctionResultObject, Matrix, Maybe } from "../../types/misc";
import {
  Granularity,
  PivotCoreDefinition,
  PivotCoreDimension,
  PivotCustomGroupedField,
  PivotDimension,
  PivotDomain,
  PivotField,
  PivotFields,
  PivotSortedColumn,
  PivotStyle,
  PivotTableCell,
} from "../../types/pivot";
import { Pivot } from "../../types/pivot_runtime";
import { SpreadsheetChildEnv } from "../../types/spreadsheet_env";
import { deepEquals, getUniqueText, isDefined } from "../misc";
import { PivotRuntimeDefinition } from "./pivot_runtime_definition";
import { pivotTimeAdapter } from "./pivot_time_adapter";

export const DEFAULT_PIVOT_STYLE: Required<PivotStyle> = {
  displayTotals: true,
  displayColumnHeaders: true,
  displayMeasuresRow: true,
  numberOfRows: Number.MAX_VALUE,
  numberOfColumns: Number.MAX_VALUE,
  tabularForm: false,
};

const AGGREGATOR_NAMES = {
  count: _t("Count"),
  count_distinct: _t("Count Distinct"),
  bool_and: _t("Boolean And"),
  bool_or: _t("Boolean Or"),
  max: _t("Maximum"),
  min: _t("Minimum"),
  avg: _t("Average"),
  sum: _t("Sum"),
};

const DEFAULT_AGGREGATORS = ["max", "min", "avg", "sum", "count_distinct", "count"];

const AGGREGATORS_BY_FIELD_TYPE = {
  integer: DEFAULT_AGGREGATORS,
  char: DEFAULT_AGGREGATORS,
  datetime: DEFAULT_AGGREGATORS,
  boolean: ["count_distinct", "count", "bool_and", "bool_or"],
};

export const AGGREGATORS = {};

for (const type in AGGREGATORS_BY_FIELD_TYPE) {
  AGGREGATORS[type] = {};
  for (const aggregator of AGGREGATORS_BY_FIELD_TYPE[type]) {
    AGGREGATORS[type][aggregator] = AGGREGATOR_NAMES[aggregator];
  }
}

type AggregatorFN = (args: Matrix<FunctionResultObject>, locale?: Locale) => FunctionResultObject;

export const AGGREGATORS_FN: Record<string, AggregatorFN | undefined> = {
  count: (args: Matrix<FunctionResultObject>) => ({
    value: countAny([args]),
    format: "0",
  }),
  count_distinct: (args: Matrix<FunctionResultObject>) => ({
    value: countUnique([args]),
    format: "0",
  }),
  bool_and: (args: Matrix<FunctionResultObject>) => ({
    value: boolAnd([args]).result,
  }),
  bool_or: (args: Matrix<FunctionResultObject>) => ({
    value: boolOr([args]).result,
  }),
  max: (args: Matrix<FunctionResultObject>, locale: Locale) => max([args], locale),
  min: (args: Matrix<FunctionResultObject>, locale: Locale) => min([args], locale),
  avg: (args: Matrix<FunctionResultObject>, locale: Locale) => ({
    value: average([args], locale),
    format: inferFormat(args),
  }),
  sum: (args: Matrix<FunctionResultObject>, locale: Locale) => ({
    value: sum([args], locale),
    format: inferFormat(args),
  }),
};

/**
 * Given an object of form {"1": {...}, "2": {...}, ...} get the maximum ID used
 * in this object
 * If the object has no keys, return 0
 *
 */
export function getMaxObjectId(o: object) {
  const keys = Object.keys(o);
  if (!keys.length) {
    return 0;
  }
  const nums = keys.map((id) => parseInt(id, 10));
  return Math.max(...nums);
}

export const ALL_PERIODS = {
  year: _t("Year"),
  quarter: _t("Quarter & Year"),
  month: _t("Month & Year"),
  week: _t("Week & Year"),
  day: _t("Day"),
  quarter_number: _t("Quarter"),
  month_number: _t("Month"),
  iso_week_number: _t("Week"),
  day_of_month: _t("Day of Month"),
  day_of_week: _t("Day of Week"),
  hour_number: _t("Hour"),
  minute_number: _t("Minute"),
  second_number: _t("Second"),
};

const DATE_FIELDS = ["date", "datetime"];

/**
 * Parse a dimension string into a pivot dimension definition.
 * e.g "create_date:month" => { name: "create_date", granularity: "month" }
 */
export function parseDimension(dimension: string): PivotCoreDimension {
  const [fieldName, granularity] = dimension.split(":");
  if (granularity) {
    return { fieldName, granularity };
  }
  return { fieldName };
}

export function isDateOrDatetimeField(field: PivotField) {
  return DATE_FIELDS.includes(field.type);
}

export function generatePivotArgs(
  formulaId: string,
  domain: PivotDomain,
  measure?: string
): string[] {
  const args: string[] = [formulaId];
  if (measure) {
    args.push(`"${measure}"`);
  }
  for (const { field, value, type } of domain) {
    if (field === "measure") {
      args.push(`"measure"`, `"${value}"`);
      continue;
    }
    const { granularity } = parseDimension(field);
    const formattedValue = toFunctionPivotValue(value, { type, granularity });
    args.push(`"${field}"`, formattedValue);
  }
  return args;
}

/**
 * Check if the fields in the domain part of
 * a pivot function are valid according to the pivot definition.
 * e.g. =PIVOT.VALUE(1,"revenue","country_id",...,"create_date:month",...,"source_id",...)
 */
export function areDomainArgsFieldsValid(dimensions: string[], definition: PivotRuntimeDefinition) {
  let argIndex = 0;
  let definitionIndex = 0;
  const cols = definition.columns.map((col) => col.nameWithGranularity);
  const rows = definition.rows.map((row) => row.nameWithGranularity);
  while (dimensions[argIndex] !== undefined && dimensions[argIndex] === rows[definitionIndex]) {
    argIndex++;
    definitionIndex++;
  }
  definitionIndex = 0;
  while (dimensions[argIndex] !== undefined && dimensions[argIndex] === cols[definitionIndex]) {
    argIndex++;
    definitionIndex++;
  }
  return dimensions.length === argIndex;
}

export function createPivotFormula(formulaId: string, cell: PivotTableCell) {
  switch (cell.type) {
    case "HEADER":
      return `=PIVOT.HEADER(${generatePivotArgs(formulaId, cell.domain).join(",")})`;
    case "VALUE":
      return `=PIVOT.VALUE(${generatePivotArgs(formulaId, cell.domain, cell.measure).join(",")})`;
    case "MEASURE_HEADER":
      return `=PIVOT.HEADER(${generatePivotArgs(formulaId, [
        ...cell.domain,
        { field: "measure", value: cell.measure, type: "char" },
      ]).join(",")})`;
  }
  return "";
}

/**
 * Parses the value defining a pivot group in a PIVOT formula
 * e.g. given the following formula PIVOT.VALUE("1", "stage_id", "42", "status", "won"),
 * the two group values are "42" and "won".
 */
export function toNormalizedPivotValue(
  dimension: Pick<PivotDimension, "type" | "displayName" | "granularity">,
  groupValue: Maybe<CellValue | FunctionResultObject>
): CellValue {
  if (groupValue === null || groupValue === "null") {
    return null;
  }
  const extractedGroupValue = typeof groupValue === "object" ? groupValue.value : groupValue;
  if (isEvaluationError(extractedGroupValue)) {
    return extractedGroupValue;
  }
  if (dimension.type === "custom") {
    return toValue(groupValue) ?? null;
  }

  const groupValueString =
    typeof groupValue === "boolean"
      ? toString(groupValue).toLocaleLowerCase()
      : toString(groupValue);
  if (groupValueString === "null") {
    return null;
  }
  if (!pivotNormalizationValueRegistry.contains(dimension.type)) {
    throw new EvaluationError(
      _t("Field %(field)s is not supported because of its type (%(type)s)", {
        field: dimension.displayName,
        type: dimension.type,
      })
    );
  }
  // represents a field which is not set (=False server side)
  if (groupValueString.toLowerCase() === "false") {
    return false;
  }
  const normalizer = pivotNormalizationValueRegistry.get(dimension.type);
  return normalizer(groupValueString, dimension.granularity);
}

function normalizeDateTime(value: CellValue, granularity: Granularity) {
  return pivotTimeAdapter(granularity ?? "month").normalizeFunctionValue(value);
}

export function toFunctionPivotValue(
  value: CellValue,
  dimension: Pick<PivotDimension, "type" | "granularity">
) {
  if (value === null) {
    return `"null"`;
  }
  if (!pivotToFunctionValueRegistry.contains(dimension.type)) {
    return `"${value}"`;
  }
  return pivotToFunctionValueRegistry.get(dimension.type)(value, dimension.granularity);
}

function toFunctionValueDateTime(value: CellValue, granularity: Granularity) {
  return pivotTimeAdapter(granularity ?? "month").toFunctionValue(value);
}

export const pivotNormalizationValueRegistry = new Registry<
  (value: string, granularity?: Granularity | string) => CellValue
>();

pivotNormalizationValueRegistry
  .add("date", normalizeDateTime)
  .add("datetime", normalizeDateTime)
  .add("integer", (value) => toNumber(value, DEFAULT_LOCALE))
  .add("boolean", (value) => toBoolean(value))
  .add("char", (value) => toString(value))
  .add("custom", (value) => value);

export const pivotToFunctionValueRegistry = new Registry<
  (value: CellValue, granularity?: string) => string
>();

pivotToFunctionValueRegistry
  .add("date", toFunctionValueDateTime)
  .add("datetime", toFunctionValueDateTime)
  .add("integer", (value: CellValue) => `${toNumber(value, DEFAULT_LOCALE)}`)
  .add("boolean", (value: CellValue) => (toBoolean(value) ? "TRUE" : "FALSE"))
  .add("char", (value: CellValue) => `"${toString(value).replace(/"/g, '\\"')}"`)
  .add("custom", (value) => (typeof value === "string" ? `"${value}"` : String(value)));

export function getFieldDisplayName(field: PivotDimension) {
  return field.displayName + (field.granularity ? ` (${ALL_PERIODS[field.granularity]})` : "");
}

export function addAlignFormatToPivotHeader(
  domain: PivotDomain,
  functionResult: FunctionResultObject
): FunctionResultObject {
  if (domain.length === 0) {
    return functionResult;
  }
  return { ...functionResult, format: (functionResult.format || "@") + "* " };
}

export function isSortedColumnValid(sortedColumn: PivotSortedColumn, pivot: Pivot): boolean {
  try {
    if (!pivot.getMeasure(sortedColumn.measure)) {
      return false;
    }

    const columns = pivot.definition.columns;
    for (let i = 0; i < sortedColumn.domain.length; i++) {
      if (columns[i].nameWithGranularity !== sortedColumn.domain[i].field) {
        return false;
      }
      const possibleValues: (CellValue | null)[] = pivot
        .getPossibleFieldValues(columns[i])
        .map((v) => v.value);
      if (
        !possibleValues.includes(sortedColumn.domain[i].value) &&
        !(sortedColumn.domain[i].value === null && possibleValues.includes(""))
      ) {
        return false;
      }
    }
    return true;
  } catch (e) {
    return false;
  }
}

export function getUniquePivotGroupName(baseName: string, field: PivotCustomGroupedField) {
  const groupNames = field.groups.map((g) => g.name);
  return getUniqueText(baseName, groupNames, {
    compute: (name, i) => `${name}${i}`,
    start: 2,
  });
}

export function getUniquePivotFieldName(baseName: string, fields: PivotFields): string {
  const namesToAvoid = Object.values(fields)
    .map((f) => [f?.name, f?.string])
    .flat()
    .filter(isDefined);
  return getUniqueText(baseName, namesToAvoid, {
    compute: (name, i) => `${name}${i}`,
    start: 2,
  });
}

export function createCustomFields(
  definition: PivotCoreDefinition,
  fields: PivotFields
): PivotFields {
  const newFields: PivotFields = {};
  for (const customField of Object.values(definition.customFields || {})) {
    const parentField = fields[customField.parentField];
    if (!parentField) {
      continue;
    }
    newFields[customField.name] = {
      type: "custom",
      isCustomField: true,
      name: customField.name,
      string: customField.name,
      customGroups: customField.groups,
      parentField: customField.parentField,
    };
  }
  return newFields;
}

export function removePivotGroupsContainingValues(
  valuesToRemove: CellValue[],
  customField: PivotCustomGroupedField
) {
  customField.groups = customField.groups.filter(
    (group) => !group.values.some((value) => valuesToRemove.includes(value))
  );
}

/**
 * Adds a new dimension to the pivot definition before a specified base dimension.
 * If the new dimension already exists, it does nothing.
 */
export function addDimensionToPivotDefinition(
  definition: PivotCoreDefinition,
  baseDimension: string,
  newDimension: string
): PivotCoreDefinition {
  const dimensions = definition.rows.some((dim) => dim.fieldName === baseDimension)
    ? definition.rows
    : definition.columns;

  const baseIndex = dimensions.findIndex((dim) => dim.fieldName === baseDimension);
  if (baseIndex === -1) {
    return definition;
  }

  if (dimensions.some((dim) => dim.fieldName === newDimension)) {
    return definition;
  }

  dimensions.splice(baseIndex, 0, { fieldName: newDimension });
  return definition;
}

export function getCustomFieldWithParentField(
  definition: PivotCoreDefinition,
  parentField: PivotField,
  fields: PivotFields
): PivotCustomGroupedField {
  return (
    Object.values(definition.customFields || {}).find(
      (field) => field.parentField === parentField.name
    ) || {
      parentField: parentField.name,
      name: getUniquePivotFieldName(parentField.string, fields),
      groups: [],
    }
  );
}

export function togglePivotCollapse(position: CellPosition, env: SpreadsheetChildEnv) {
  const pivotCell = env.model.getters.getPivotCellFromPosition(position);
  const pivotId = env.model.getters.getPivotIdFromPosition(position);
  if (!pivotId || pivotCell.type !== "HEADER") {
    return;
  }
  const definition = env.model.getters.getPivotCoreDefinition(pivotId);

  const collapsedDomains = definition.collapsedDomains?.[pivotCell.dimension]
    ? [...definition.collapsedDomains[pivotCell.dimension]]
    : [];
  const index = collapsedDomains.findIndex((domain) => deepEquals(domain, pivotCell.domain));
  if (index !== -1) {
    collapsedDomains.splice(index, 1);
  } else {
    collapsedDomains.push(pivotCell.domain);
  }

  const newDomains = definition.collapsedDomains
    ? { ...definition.collapsedDomains }
    : { COL: [], ROW: [] };
  newDomains[pivotCell.dimension] = collapsedDomains;
  env.model.dispatch("UPDATE_PIVOT", {
    pivotId,
    pivot: { ...definition, collapsedDomains: newDomains },
  });
}

export function getPivotStyleFromFnArgs(
  definition: PivotCoreDefinition,
  rowCountArg: Maybe<FunctionResultObject | CellValue>,
  includeTotalArg: Maybe<FunctionResultObject | CellValue>,
  includeColumnHeadersArg: Maybe<FunctionResultObject | CellValue>,
  columnCountArg: Maybe<FunctionResultObject | CellValue>,
  includeMeasuresRowArg: Maybe<FunctionResultObject | CellValue>,
  tabularFormArg: Maybe<FunctionResultObject | CellValue>,
  locale: Locale
): Required<PivotStyle> {
  const style = definition.style;

  const numberOfRows =
    rowCountArg !== undefined
      ? toNumber(rowCountArg, locale)
      : style?.numberOfRows ?? DEFAULT_PIVOT_STYLE.numberOfRows;
  const numberOfColumns =
    columnCountArg !== undefined
      ? toNumber(columnCountArg, locale)
      : style?.numberOfColumns ?? DEFAULT_PIVOT_STYLE.numberOfColumns;

  const displayTotals =
    includeTotalArg !== undefined
      ? toBoolean(includeTotalArg)
      : style?.displayTotals ?? DEFAULT_PIVOT_STYLE.displayTotals;
  const displayColumnHeaders =
    includeColumnHeadersArg !== undefined
      ? toBoolean(includeColumnHeadersArg)
      : style?.displayColumnHeaders ?? DEFAULT_PIVOT_STYLE.displayColumnHeaders;
  const displayMeasuresRow =
    includeMeasuresRowArg !== undefined
      ? toBoolean(includeMeasuresRowArg)
      : style?.displayMeasuresRow ?? DEFAULT_PIVOT_STYLE.displayMeasuresRow;

  const tabularForm =
    tabularFormArg !== undefined
      ? toBoolean(tabularFormArg)
      : style?.tabularForm ?? DEFAULT_PIVOT_STYLE.tabularForm;

  return {
    numberOfRows,
    numberOfColumns,
    displayTotals,
    displayColumnHeaders,
    displayMeasuresRow,
    tabularForm,
  };
}
