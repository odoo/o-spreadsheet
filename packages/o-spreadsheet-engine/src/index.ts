export { FunctionCodeBuilder, Scope, type FunctionCode } from "./formulas/code_builder";
export {
  OPERATOR_MAP,
  UNARY_OPERATOR_MAP,
  compile,
  compileTokens,
  functionCache,
} from "./formulas/compiler";
export {
  OP_PRIORITY,
  convertAstNodes,
  iterateAstNodes,
  mapAst,
  parse,
  parseTokens,
} from "./formulas/parser";
export type {
  AST,
  ASTFuncall,
  ASTOperation,
  ASTString,
  ASTSymbol,
  ASTUnaryOperation,
} from "./formulas/parser";
export { rangeTokenize } from "./formulas/range_tokenizer";
export { POSTFIX_UNARY_OPERATORS, tokenize } from "./formulas/tokenizer";
export type { Token } from "./formulas/tokenizer";
export { FunctionRegistry } from "./functions/function_registry";
export { categories } from "./functions/functionRegistry";
export * from "./helpers";
export { Model } from "./model";
export { BasePlugin } from "./plugins/base_plugin";
export { Registry } from "./registry";
export { StateObserver, type StateObserverChange } from "./state_observer";
export * from "./translation";
export { CellValue } from "./types/cells";
export { SpreadsheetClipboardData } from "./types/clipboard";
export { CoreGetters, PluginGetters } from "./types/coreGetters";
export * from "./types/errors";
export { Format } from "./types/format";
export * from "./types/history";
export * from "./types/locale";
export * from "./types/misc";
export * from "./types/validator";
export const __info__ = {};
