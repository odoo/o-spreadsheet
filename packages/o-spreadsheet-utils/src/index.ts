export * from "./translation";
export * from "./uuid";
