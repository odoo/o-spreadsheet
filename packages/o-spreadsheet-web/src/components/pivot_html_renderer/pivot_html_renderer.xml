<templates>
  <t t-name="o_spreadsheet.PivotHTMLRenderer">
    <div class="o_pivot_html_renderer">
      <Checkbox
        name="'missing_values'"
        label.translate="Display missing cells only"
        value="state.showMissingValuesOnly"
        onChange.bind="(value) => this.state.showMissingValuesOnly = value"
        className="'m-2'"
      />
      <t t-set="tableData" t-value="getTableData()"/>
      <table
        class="o_pivot_html_renderer"
        t-if="tableData.values.length > 0 or tableData.rows.length > 0">
        <tr t-foreach="tableData.columns" t-as="row" t-key="row_index">
          <t t-if="row_index === 0">
            <th t-att-rowspan="tableData.columns.length"/>
          </t>
          <t t-foreach="row" t-as="cell" t-key="cell_index">
            <th
              t-att-colspan="cell.span"
              t-att-style="cell.style"
              t-att-class="{ o_missing_value: cell.isMissing }"
              t-on-click="() => props.onCellClicked(cell.formula)">
              <t t-esc="cell.value"/>
            </th>
          </t>
        </tr>
        <t t-foreach="tableData.rows" t-as="row" t-key="row_index">
          <tr>
            <th
              t-att-style="row.style"
              t-att-class="{ o_missing_value: row.isMissing }"
              t-on-click="() => props.onCellClicked(row.formula)">
              <t t-esc="row.value"/>
            </th>
            <t t-foreach="tableData.values" t-as="col" t-key="col_index">
              <td
                t-att-class="{ o_missing_value: col[row_index].isMissing }"
                t-on-click="() => props.onCellClicked(col[row_index].formula)">
                <t t-esc="col[row_index].value"/>
              </td>
            </t>
          </tr>
        </t>
      </table>
      <div class="alert alert-info" t-else="1">This pivot has no cell missing on this sheet</div>
    </div>
  </t>
</templates>
