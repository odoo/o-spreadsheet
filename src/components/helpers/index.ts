export * from "@odoo/o-spreadsheet-engine/components/helpers/css";
