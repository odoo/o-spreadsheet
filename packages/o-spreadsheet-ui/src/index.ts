import * as Engine from "@odoo/o-spreadsheet-engine";

export function salut() {
  console.log("Salut !");
}

export function createSpreadsheetModel() {
  return Engine.SpreadsheetModel();
}