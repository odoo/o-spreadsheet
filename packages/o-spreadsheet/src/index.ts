import * as UI from "@odoo/o-spreadsheet-ui";


export { UI };