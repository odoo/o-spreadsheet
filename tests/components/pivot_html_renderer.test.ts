import { UID } from "@odoo/o-spreadsheet-engine";
import { Model } from "../../src";
import { PivotHTMLRenderer } from "../../src/components/pivot_html_renderer/pivot_html_renderer";
import { createSheet } from "../test_helpers/commands_helpers";
import { click } from "../test_helpers/dom_helper";
import { createModelFromGrid, mountComponent } from "../test_helpers/helpers";
import { addPivot } from "../test_helpers/pivot_helpers";

let fixture: HTMLElement;

async function mountPivotHtmlRenderer(
  model: Model,
  pivotId: UID,
  onCellClicked: PivotHTMLRenderer["props"]["onCellClicked"] = () => {}
) {
  const props = {
    pivotId,
    onCellClicked,
  };
  model.dispatch("PIVOT_START_PRESENCE_TRACKING", { pivotId });
  model.dispatch("EVALUATE_CELLS");
  ({ fixture } = await mountComponent(PivotHTMLRenderer, { env: { model }, props }));
}

describe("Pivot HTML Renderer", () => {
  test("Rendering a simple pivot table", async () => {
    // prettier-ignore
    const grid = {
      A1: "Name",  B1: "Age", C1: "Score",
      A2: "Alice", B2: "25",  C2: "90",
      A3: "Bob",   B3: "30",  C3: "85",
    };
    const model = createModelFromGrid(grid);
    addPivot(model, "A1:C3", {
      rows: [{ fieldName: "Name" }],
      measures: [{ id: "Score", fieldName: "Score", aggregator: "count" }],
    });

    await mountPivotHtmlRenderer(model, model.getters.getPivotIds()[0]);
    expect(fixture).toMatchSnapshot();
  });

  test("Pivot with all formula on sheet", async () => {
    // prettier-ignore
    const grid = {
      A1: "Name",                            B1: "Age",                                    C1: "Score",
      A2: "Alice",                           B2: "25",                                     C2: "90",
      A3: "Bob",                             B3: "30",                                     C3: "85",

      A5: "",                                B5: "=PIVOT.HEADER(1)",
      A6: "",                                B6: `=PIVOT.HEADER(1,"measure","Score")`,
      A7: `=PIVOT.HEADER(1,"Name","Alice")`, B7: `=PIVOT.VALUE(1,"Score","Name","Alice")`,
      A8: `=PIVOT.HEADER(1,"Name","Bob")`,   B8: `=PIVOT.VALUE(1,"Score","Name","Bob")`,
      A9: "=PIVOT.HEADER(1)",                B9: `=PIVOT.VALUE(1,"Score")`,
    };
    const model = createModelFromGrid(grid);
    addPivot(model, "A1:C3", {
      rows: [{ fieldName: "Name" }],
      measures: [{ id: "Score", fieldName: "Score", aggregator: "count" }],
    });
    await mountPivotHtmlRenderer(model, model.getters.getPivotIds()[0]);
    expect(fixture.querySelectorAll(".o_missing_value")).toHaveLength(0);
    await click(fixture, "input[type=checkbox]");
    expect(fixture.querySelector("table")).toBeNull();
    expect(fixture.querySelector(".alert-info")?.innerHTML).toEqual(
      "This pivot has no cell missing on this sheet"
    );
  });

  test("Pivot with all formula on another sheet", async () => {
    // prettier-ignore
    const grid = {
      A1: "Name",                            B1: "Age",                                    C1: "Score",
      A2: "Alice",                           B2: "25",                                     C2: "90",
      A3: "Bob",                             B3: "30",                                     C3: "85",

      A5: "",                                B5: "=PIVOT.HEADER(1)",
      A6: "",                                B6: `=PIVOT.HEADER(1,"measure","Score")`,
      A7: `=PIVOT.HEADER(1,"Name","Alice")`, B7: `=PIVOT.VALUE(1,"Score","Name","Alice")`,
      A8: `=PIVOT.HEADER(1,"Name","Bob")`,   B8: `=PIVOT.VALUE(1,"Score","Name","Bob")`,
      A9: "=PIVOT.HEADER(1)",                B9: `=PIVOT.VALUE(1,"Score")`,
    };
    const model = createModelFromGrid(grid);
    addPivot(model, "A1:C3", {
      rows: [{ fieldName: "Name" }],
      measures: [{ id: "Score", fieldName: "Score", aggregator: "count" }],
    });
    createSheet(model, { activate: true });
    await mountPivotHtmlRenderer(model, model.getters.getPivotIds()[0]);
    expect(fixture.querySelectorAll(".o_missing_value")).toHaveLength(8);
  });
});
