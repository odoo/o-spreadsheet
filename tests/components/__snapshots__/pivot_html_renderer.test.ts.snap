// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pivot HTML Renderer Rendering a simple pivot table 1`] = `
<div>
  <div
    class="o_pivot_html_renderer"
  >
    <label
      class="o-checkbox d-flex align-items-center m-2"
    >
      <input
        class="me-2"
        name="missing_values"
        type="checkbox"
      />
      Display missing cells only
      
    </label>
    
    <table
      class="o_pivot_html_renderer"
    >
      <tr>
        <th
          rowspan="2"
        />
        
        <th
          class="o_missing_value"
          colspan="1"
        >
          Total
        </th>
        
        
      </tr>
      <tr>
        
        <th
          class="o_missing_value"
          colspan="1"
        >
          Score
        </th>
        
        
      </tr>
      
      
      <tr>
        <th
          class="o_missing_value"
        >
          Alice
        </th>
        <td
          class="o_missing_value"
        >
          1
        </td>
        
        
      </tr>
      <tr>
        <th
          class="o_missing_value"
        >
          Bob
        </th>
        <td
          class="o_missing_value"
        >
          1
        </td>
        
        
      </tr>
      <tr>
        <th
          class="o_missing_value"
        >
          Total
        </th>
        <td
          class="o_missing_value"
        >
          2
        </td>
        
        
      </tr>
      
      
    </table>
    
    
  </div>
</div>
`;
